File: pep249_dbapi/tests/integ/test_put_get_overwrite.py
from .utils_put_get import (
    as_file_uri,
    create_temporary_stage,
    PUT_ROW_SOURCE_IDX,
    PUT_ROW_STATUS_IDX,
)
from ..utils import shared_test_data_dir


def test_put_overwrite_true(cursor):
    stage_name = create_temporary_stage(cursor, "PYTEST_STAGE_OVERWRITE_TRUE")
    filename, original = original_test_file()
    _, updated = updated_test_file()

    # Upload the file to the stage
    cursor.execute(
        f"PUT 'file://{as_file_uri(original)}' @{stage_name}"
    )

    # Verify that the file was uploaded
    row = cursor.fetchone()
    assert row[PUT_ROW_SOURCE_IDX] == filename
    assert row[PUT_ROW_STATUS_IDX] == "UPLOADED"

    # Upload again with changed content and OVERWRITE=TRUE
    cursor.execute(
        f"PUT 'file://{as_file_uri(updated)}' @{stage_name} OVERWRITE=TRUE"
    )

    # Verify that the file was uploaded
    row = cursor.fetchone()
    assert row[PUT_ROW_SOURCE_IDX] == filename
    assert row[PUT_ROW_STATUS_IDX] == "UPLOADED"

    # Verify that the content was updated
    cursor.execute(f"SELECT $1, $2, $3 FROM @{stage_name}")
    row = cursor.fetchone()
    assert row == ("updated", "test", "data")


def test_put_overwrite_false(cursor):
    stage_name = create_temporary_stage(cursor, "PYTEST_STAGE_OVERWRITE_FALSE")
    filename, original = original_test_file()
    _, updated = updated_test_file()

    # Upload the file to the stage
    cursor.execute(
        f"PUT 'file://{as_file_uri(original)}' @{stage_name}"
    )

    # Verify that the file was uploaded
    row = cursor.fetchone()
    assert row[PUT_ROW_SOURCE_IDX] == filename
    assert row[PUT_ROW_STATUS_IDX] == "UPLOADED"

    # Try to upload changed content with OVERWRITE=FALSE
    cursor.execute(
        f"PUT 'file://{as_file_uri(updated)}' @{stage_name} OVERWRITE=FALSE"
    )

    # Verify that the file was not uploaded
    row = cursor.fetchone()
    assert row[PUT_ROW_SOURCE_IDX] == filename
    assert row[PUT_ROW_STATUS_IDX] == "SKIPPED"

    # Verify original content remains
    cursor.execute(f"SELECT $1, $2, $3 FROM @{stage_name}")
    row = cursor.fetchone()
    assert row == ("original", "test", "data")


def original_test_file():
    return (
        "test_data.csv",
        shared_test_data_dir() / "overwrite" / "original" / "test_data.csv",
    )


def updated_test_file():
    return (
        "test_data.csv",
        shared_test_data_dir() / "overwrite" / "updated" / "test_data.csv",
    )

File: odbc_tests/tests/put_get_tests/test_put_get_overwrite.cpp
#include <algorithm>
#include <filesystem>
#include <fstream>
#include <random>
#include <string>
#include <tuple>
#include <vector>

#include <catch2/catch_test_macros.hpp>

#include "Connection.hpp"
#include "get_data.hpp"
#include "put_get_utils.hpp"

namespace fs = std::filesystem;
using namespace pg_utils;

static std::pair<std::string, fs::path> original_test_file() {
  return {"test_data.csv", shared_test_data_dir() / "overwrite" / "original" / "test_data.csv"};
}

static std::pair<std::string, fs::path> updated_test_file() {
  return {"test_data.csv", shared_test_data_dir() / "overwrite" / "updated" / "test_data.csv"};
}

TEST_CASE("PUT overwrite true", "[put_get][odbc]") {
  Connection conn;
  const std::string stage = pg_utils::create_stage(conn, "ODBCTST_OVERWRITE_TRUE");
  auto [filename, original] = original_test_file();
  auto [_, updated] = updated_test_file();

  {
    // Upload file
    auto stmt = conn.execute_fetch("PUT 'file://" + as_file_uri(original) + "' @" + stage);
    std::string src = get_data<SQL_C_CHAR>(stmt, PUT_ROW_SOURCE_IDX);
    std::string status = get_data<SQL_C_CHAR>(stmt, PUT_ROW_STATUS_IDX);
    CHECK(src == filename);
    CHECK(status == "UPLOADED");
  }

  {
    // Overwrite existing file
    auto stmt = conn.execute_fetch("PUT 'file://" + as_file_uri(updated) + "' @" + stage +
                                   " OVERWRITE=TRUE");
    std::string src = get_data<SQL_C_CHAR>(stmt, PUT_ROW_SOURCE_IDX);
    std::string status = get_data<SQL_C_CHAR>(stmt, PUT_ROW_STATUS_IDX);
    CHECK(src == filename);
    CHECK(status == "UPLOADED");
  }

  // Verify the data matches what we uploaded
  {
    auto stmt = conn.execute_fetch("SELECT $1, $2, $3 FROM @" + stage);
    std::string c1 = get_data<SQL_C_CHAR>(stmt, 1);
    std::string c2 = get_data<SQL_C_CHAR>(stmt, 2);
    std::string c3 = get_data<SQL_C_CHAR>(stmt, 3);
    CHECK(c1 == "updated");
    CHECK(c2 == "test");
    CHECK(c3 == "data");
  }
}

TEST_CASE("PUT overwrite false", "[put_get][odbc]") {
  Connection conn;
  const std::string stage = pg_utils::create_stage(conn, "ODBCTST_OVERWRITE_FALSE");
  auto [filename, original] = original_test_file();
  auto [_, updated] = updated_test_file();

  {
    // Upload file
    auto stmt = conn.execute_fetch("PUT 'file://" + as_file_uri(original) + "' @" + stage);
    std::string src = get_data<SQL_C_CHAR>(stmt, PUT_ROW_SOURCE_IDX);
    std::string status = get_data<SQL_C_CHAR>(stmt, PUT_ROW_STATUS_IDX);
    CHECK(src == filename);
    CHECK(status == "UPLOADED");
  }

  {
    // Attempt overwrite with OVERWRITE=FALSE -> should be skipped
    auto stmt = conn.execute_fetch("PUT 'file://" + as_file_uri(updated) + "' @" + stage +
                                   " OVERWRITE=FALSE");
    std::string src = get_data<SQL_C_CHAR>(stmt, PUT_ROW_SOURCE_IDX);
    std::string status = get_data<SQL_C_CHAR>(stmt, PUT_ROW_STATUS_IDX);
    CHECK(src == filename);
    CHECK(status == "SKIPPED");
  }

  // Verify the data matches what we uploaded
  {
    auto stmt = conn.execute_fetch("SELECT $1, $2, $3 FROM @" + stage);
    std::string c1 = get_data<SQL_C_CHAR>(stmt, 1);
    std::string c2 = get_data<SQL_C_CHAR>(stmt, 2);
    std::string c3 = get_data<SQL_C_CHAR>(stmt, 3);
    CHECK(c1 == "original");
    CHECK(c2 == "test");
    CHECK(c3 == "data");
  }
}
